### ISSUE TITLE
Scala API

### ISSUE THREAD

**Issue by eaplatanios** [tensorflow/tensorflow#10089]
Hi,

I have open sourced a Scala API for TensorFlow that contains much more complete functionality than the Java API, [here](https://github.com/eaplatanios/tensorflow_scala). The README file in the repository contains information on what features are supported etc. I would really appreciate some feedback from interested parties in the community on the library.

One main limitation is that the library does not yet support while loops in the graph. This is due to being unable to implement gradient backpropagation using the current C API. @skye do you have any suggestions on how to proceed with respect to that? It would be really useful for implementing RNNs.

Other cool stuff (such as fetching arbitrarily structured data from sessions -- e.g., lists of tensors, or maps of tensors, etc.) are supported and are for the most part strongly typed.

Cheers,
Anthony

**Comment by skye** [comment]
Super cool Anthony! This looks really great. FYI @asimshankar

Regarding while loops, I still need to reimplement the backprop logic in the C API as we discussed in #9150. Given that we're both depending on this (I'm currently switching the Python API to use the C API), I'll start work on this ASAP. I can't make any promises as to when it will land, but hopefully within the next month depending on the complexity and what else comes up (I'll also be on vacation next week).

**Comment by eaplatanios** [comment]
@skye Thanks! :)

Regarding the gradients, I'm currently not using the provided C API gradients, because it's missing a lot of them (my library still supports it but the default one is my implementation). I have instead implemented back-propagation in the Scala API (essentially porting the Python API code). In this case it would be great for your implementation to allow me to implement back-propagation over while loops in the Scala side, or to allow registering gradient functions through the C API (although I feel this is not easy to do). Do you have any thoughts on this?

P.S. I'm actually also going on vacation in a couple weeks! Have a great time wherever you're going! :))

**Comment by jart** [comment]
Thank you @eaplatanios for bringing your work to the attention of the community. We appreciate what you're doing. We hope the Scala community will find TensorFlow helpful.

Here's one way we might be able to support you. You could send us a pull request updating our documentation to point to your project. For example, the [Which client languages are supported in TensorFlow?](https://www.tensorflow.org/programmers_guide/faq) section in our Programmer's Guide. You could add a sentence that says something like, "There is also an independently maintained interface for [Scala](https://github.com/eaplatanios/tensorflow_scala)." 

Any thoughts @martinwicke?

CC: @jdegoes

**Comment by asimshankar** [comment]
@eaplatanios : This is wonderful. If you'd like, I'd be happy to also link to this from https://www.tensorflow.org/api_docs/ like we do for some other language bindings.

**Comment by jdegoes** [comment]
@eaplatanios Nice work! I'll offer my assistance on the API for this library if you would like.

**Comment by skye** [comment]
@eaplatanios I'll have to think about the while loop backprop more. We have the same problem porting the Python while_loop to use the C API, since the C gradients aren't finished. @asimshankar and I had discussed the C loop backprop code calling back into the Python runtime to create the individual gradients, so maybe we can make that mechanism general enough to also use in the Scala API. I'll let you know when I come up with something more concrete, and let me know if you have any further thoughts.

**Comment by eaplatanios** [comment]
@jart Thanks! That sounds good. I can do that.

@asimshankar Thanks! That would be great! :)

@jdegoes Thanks! That would actually be helpful as there are some parts of the library I'm not happy with and would like to make more elegant. We can talk about it whenever you want and I can tell you my thoughts. :)

@skye C backprop code calling back into the Python runtime to create the individual gradients sounds like what I was referring to although it's hard to make generic. One possible idea would be to allow creation of functions (i.e., subgraphs) in the C API, and then define control flow operations in terms of those functions. Then, a single function for returning the gradient of a while block should be sufficient for users that implement backprop in other languages. Or something along those lines.

**Comment by eaplatanios** [comment]
@skye How about the following proposal?

I have not really looked much at how functions (i.e., subgraphs) are defined in graphs, but intuitively I think the following solution makes sense.

Currently, there exist gradient registration functions in the Python API and in the C++ API, as well as in my Scala API. Furthermore, the gradient back-propagation code is also defined in all three places. If we change the gradient registration functions to instead return a TF function (i.e., a sub-graph) that has the appropriate inputs and outputs, then that would make these sub-graphs "transferrable" between other language APIs and the C API.

This should allow for the back-propagation code to only exist in the C++ and then we could have C API functions for registering op gradients, which would only need the op type name and the gradient function (i.e., sub-graph in this case) as arguments. This will allow registering gradient functions for yet unsupported gradients in the C++ side, while also avoiding replication of the gradient back-propagation code.

I hope my brief description makes sense.

**Comment by asimshankar** [comment]
@eaplatanios : Use of functions would be a great way to try and make gradients defined in any one language available to others and is something we are interested in. However, the timeline for that isn't clear yet (as a first step there are a few things we have to do before TF functions themselves are made part of the public API, and even after that there are some missing features in TF functions that are needed before all gradient functions can be expressed). What @skye suggested (having the C API call back into Python for the gradients) is a means of decoupling the transition to using the C API for graph construction in Python from working out all of those issues.

**Comment by eaplatanios** [comment]
@asimshankar I see. I can see how this may work with Python, but how would you go about having the C API call back into Java or Scala functions for the gradients. Is there a simple way to do that with JNI? Or is there some more generic way that would allow you to do that with any language?

One way I see for doing this would be to have the C API expose a function interface that users of the API can implement and which is similar to the gradient function interface in the Python API or my Scala API. Then, users can implement that by having it lookup a functions registry in their language implementation that would return the gradients if there is a corresponding function registered. The C API function would be called whenever the C API needs gradients for some op. I also think it might be better for that function to be called first (for now), before looking for C++ gradient functions, in order to retain consistency in how the gradients are named in the constructed graph. Does that sound like something that's possible with the C API? I'm not very familiar with C to know if that's easy to do or not.

Yet another way may be similarly to how tensor deallocators are specified. The users could possible register gradient functions this way. In this case, they would only need to register a single function that delegates the gradient lookup to their own language implementation, which constructs the gradients and returns the result.

**Comment by Moriadry** [comment]
It's so cool man, thank you for your giving:)

**Comment by eaplatanios** [comment]
@asimshankar I created a pull request to link to my repository from some of the documentation pages, but I could not figure out how to link to it from [https://www.tensorflow.org/api_docs/](https://www.tensorflow.org/api_docs/) as you suggested. Could you please help me with that?

@skye Is there any update on the support for gradients over while loops, etc., that we discussed before? Thanks! :)

@Moriadry Thanks! :)

**Comment by asimshankar** [comment]
@eaplatanios : I believe that particular landing page isn't checked into github, but I will take care of adding that before the next release. Thanks!

**Comment by skye** [comment]
@eaplatanios: I've been working on the framework for generating the backprop graph of while loops, but haven't committed anything yet. I'll post here as I make progress, and once the general framework is in we'll have to address how to generate unimplemented C++ op gradients.

**Comment by ysihaoy** [comment]
That's really good, thanks.

We have community version of Scala GRPC and now we have Scala API for tensorflow.

Hopefully Google could add those as official support.

**Comment by martinwicke** [comment]
I doubt we'd add official support, given our Scala bench isn't very deep.
However, we'd be excited to link to Scala bindings from tensorflow.org.

**Comment by eaplatanios** [comment]
@martinwicke That is totally understandable! Do you mean linking on [this page](https://www.tensorflow.org/api_docs/) for the 1.3 release? That would be great! :)

**Comment by eaplatanios** [comment]
@skye That sounds great! I look forward to the updates and I can certainly provide some feedback on how to to generate unimplemented C++ op gradients when time comes. :)

**Comment by martinwicke** [comment]
Yes, I believe Asim already mentioned the possibility earlier.

**Comment by eaplatanios** [comment]
Ok that sounds great. You're right, I'm sorry I had forgotten about the previous message. In this case, let's go ahead and do this. :)

**Comment by martinwicke** [comment]
Please do send a PR. I believe all this is in GitHub. @wolffg can you
confirm?

**Comment by eaplatanios** [comment]
I tried to find that page before but couldn't. @asim mentioned before that that page may not be in github.

**Comment by asimshankar** [comment]
@martinwicke @eaplatanios : Yeah that page isn't in github, it's generated as part of the script that uploads to www.tensorflow.org. I've made a change there, so it should go out in the next website push

**Comment by eaplatanios** [comment]
@asimshankar Great, thanks Asim! :)

**Comment by debasish83** [comment]
@eaplatanios @asimshankar are there plans to combine the jni library in tensorflow_scala with the java interface ? tensorflow_scala exposes the Server interfaces which will help in implementing distributed training and inference using parameter servers deployed in Spark clusters....

**Comment by eaplatanios** [comment]
@debasish83 It's very easy to share the work I have done for my JNI bindings and I would be happy to collaborate and find a way to merge it into the main TensorFlow repository. However, most of the setup/management for distributed TF in my library is implemented on the Scala side. The C API does not expose much for that as a lot of it is only implemented in the Python API. However, it's easy to create a Java API to my Scala API if you're up for that. You would just have to work around my use of implicits in the Scala side which should not be too bad for distributed training. However, it might get tricky if you try to create a Java interface for my learn API (i.e., estimators, datasets, etc.) as I use a lot of implicits for strong type checking there. Note that the best way to perform distributed training currently in my library is through the learn API I provide (similar to the Python API except for the strong typing and some other features in which we diverge). I'd be happy to provide support if you're interested in working on something like that. :)

**Comment by eaplatanios** [comment]
@skye @asimshankar I have also finally added support for while loops, back propagation, and RNNs, by reimplementing pretty much all of the Python API functionality and adding some strong typing on top of it (with respect to loop variables, etc), and so I guess you can close this issue if you want as there isn't any remaining issue I can see. :)

**Comment by debasish83** [comment]
@eaplatanios I am fine to use scala API...like MXNet it will be useful to introduce a scala package to support distributed learning with spark/flink and model inference can continue using java/scala API...I will provide further feedbacks on https://github.com/eaplatanios/tensorflow_scala...

**Comment by eaplatanios** [comment]
@debasish83 Sounds good! I'd really appreciate your feedback. :) I'm currently in the process of adding a sequence-to-sequence modeling package as well as a neural machine translation module. I will be working for a paper deadline until December 15th and traveling, but I'll try to be as responsive as possible. :)

**Comment by skye** [comment]
Wow awesome @eaplatanios! I'll go ahead and close this issue, but feel free to continue the current discussion here.
